File: frontend/src/types/scanner.ts
export type ScanResult = {
  symbol: string;
  company_name: string | null;
  price: number | null;
  change_percent: number | null;
  predicted_price: number | null;
  predicted_change_percent: number | null;
  stockpilot_score: number | null;
  volume: number | null;
  market_cap: number | null;
  pe_ratio: number | null;
  rsi_14: number | null;
  technical_score: number | null;
  momentum_score: number | null;
  risk_score: number | null;
  overall_rating: string | null;
  recommendation: string | null;
  analyst_sentiment: string | null;
  ai_reasoning: string | null;
  confidence_score: number | null;
  sector: string | null;
  provider: string | null;
};

export type ScannerResponse = {
  results: ScanResult[];
  scanned_at: string;
  universe_size: number;
  filters_applied: Record<string, string | number | boolean>;
  disclaimer: string;
};

export type ChatMessage = {
  id?: string;
  role: "user" | "assistant";
  content: string;
};

export type NewsStockMention = {
  symbol: string;
  company_name: string | null;
  price: number | null;
  change_percent: number | null;
};

export type NewsArticle = {
  title: string;
  source: string | null;
  url: string | null;
  published_at: string | null;
  summary: string | null;
  image_url: string | null;
  symbols: string[];
  related_stocks: NewsStockMention[];
};

export type FallenGiantSource = {
  title: string;
  source: string | null;
  url: string | null;
  published_at: string | null;
};

export type FallenGiantCandidate = {
  symbol: string;
  company_name: string | null;
  market_cap: number | null;
  current_price: number | null;
  pre_catalyst_price: number | null;
  lowest_price_after_catalyst: number | null;
  decline_percent: number | null;
  recovered_percent: number | null;
  catalyst: string | null;
  catalyst_type: string;
  catalyst_date: string | null;
  fundamental_health_score: number | null;
  valuation_score: number | null;
  recovery_score: number | null;
  risk_score: number | null;
  catalyst_clarity_score: number | null;
  price_dislocation_score: number | null;
  fallen_giants_score: number | null;
  why_it_fell: string | null;
  why_it_could_recover: string | null;
  why_it_might_not_recover: string | null;
  recovery_confirmations: string[];
  sources: FallenGiantSource[];
  sector: string | null;
  relative_to_spy_decline: number | null;
  days_since_catalyst: number | null;
  selloff_days: number | null;
  provider: string | null;
  data_warnings: string[];
};

export type FallenGiantsResponse = {
  results: FallenGiantCandidate[];
  scanned_at: string;
  universe_size: number;
  filters_applied: Record<string, string | number | boolean>;
  disclaimer: string;
};

File: backend/app/schemas/scanner.py
from datetime import datetime, timezone

from pydantic import BaseModel, Field


class ScanResult(BaseModel):
    symbol: str
    company_name: str | None = None
    price: float | None = None
    change_percent: float | None = None
    predicted_price: float | None = None
    predicted_change_percent: float | None = None
    stockpilot_score: float | None = None
    volume: int | None = None
    market_cap: float | None = None
    pe_ratio: float | None = None
    rsi_14: float | None = None
    technical_score: float | None = None
    momentum_score: float | None = None
    risk_score: float | None = None
    overall_rating: str | None = None
    recommendation: str | None = None
    analyst_sentiment: str | None = None
    ai_reasoning: str | None = None
    confidence_score: float | None = None
    sector: str | None = None
    provider: str | None = None


class ScannerResponse(BaseModel):
    results: list[ScanResult]
    scanned_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    universe_size: int
    filters_applied: dict[str, str | float | int | bool] = Field(default_factory=dict)
    disclaimer: str = (
        "AI-ranked opportunities from live market data — not financial advice."
    )


class ScreenerFilters(BaseModel):
    min_rsi: float | None = Field(None, ge=0, le=100)
    max_rsi: float | None = Field(None, ge=0, le=100)
    min_technical_score: float | None = Field(None, ge=0, le=100)
    min_momentum_score: float | None = Field(None, ge=0, le=100)
    sector: str | None = None
    min_price: float | None = Field(None, ge=0)
    max_price: float | None = Field(None, ge=0)
    investment_horizon: str | None = None
    risk_level: str | None = None
    query: str | None = Field(None, description="Symbol, company name, or sector search")
    limit: int = Field(50, ge=1, le=100)


class ChatMessage(BaseModel):
    role: str
    content: str


class ChatContext(BaseModel):
    page: str | None = None
    symbol: str | None = None


class ChatRequest(BaseModel):
    message: str = Field(..., min_length=1, max_length=4000)
    history: list[ChatMessage] = Field(default_factory=list)
    context: ChatContext | None = None


class ChatResponse(BaseModel):
    reply: str
    disclaimer: str = "AI assistant for research only — not financial advice. Verify all figures."
